text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+I'm not sure how to address this. It might require the oracle to be able to
+operate in UTF16 mode.
+
+Where no return value is specified, the call returns success or failure (with a
+reason).
+
+The philosophy of the API is that most functions should only take a single call,
+as opposed to making each function as minimal and orthogonal as possible. This
+is because IPC can be slow and response time is important for IDEs.
+
+
+**Projects**
+
+Note that the oracle stores no metadata about a project.
+
+*init project*
+
+Takes a project name, returns an id string (something close to the project's name).
+
+*delete project*
+
+Takes a project id.
+
+*list projects*
+
+Takes nothing, returns a list of project ids.
+
+Each of the remaining calls takes a project identifier.
+
+
+**Update**
+
+See section on input data format below.
+
+*update*
+
+Takes input data (actual source code rather than spans since we cannot assume
+the user has saved the file) and a list of spans to invalidate.  Where there are
+no invalidated spans, the update call adds data (which will cause an error if
+there are conflicts). Where there is no input data, update just invalidates.
+
+We might want to allow some shortcuts to invalidate an entire file or
+recursively invalidate a directory.
+
+
+**Description**
+
+*get definition*
+
+Takes a span, returns all 'definitions and declarations' for the identifier
+covered by the span. Can return an error if the span does not cover exactly one
+identifier or the oracle has no data for an identifier.
+
+The returned data is a list of 'defintion' data. That data includes the span for
+the item, any documentation for the item, a code snippet for the item,
+optionally a type for the item, and one or more kinds of definition (e.g.,
+'variable definition', 'field definition', 'function declaration').
+
+*get references*
+
+Takes a span, returns a list of reference data (or an error). Each datum
+consists of the span of the reference and a code snippet.
+
+*get docs*
+
+Takes a span, returns the same data as *get definition* but limited to doc strings.
+
+*get type*
+
+Takes a span, returns the same data as *get definition* but limited to type information.
+
+Question: are these useful/necessary? Or should users just call *get definition*?
+
+*search for identifier*
+
+Takes a search string or an id, and a struct of search parameters including case
+sensitivity, and the kind of items to search (e.g., functions, traits, all
+items). Returns a list of spans and code snippets.
+
+
+**Code completion**
+
+*get suggestions*
+
+Takes a span (note that this span could be empty, e.g, for `foo.` we would use
+the empty span which starts after the `.`; for `foo.b` we would use the span for
+`b`), and returns a list of suggestions (is this useful? Is there any difference
+from just using the caret position?). Each suggestion consists of the text for
+completion plus the same information as returned for the *get definition* call.
+
+
```

Comment:
Not sure about the specifics. With a Qt project, this Locator feature matches either a symbol name, or a word that appears inside a title of the generated html doc (they are structured docs). This is clearly not a critical feature, it may not belong to an initial RFC.